enzyme/test/Integration/ReverseMode/invsqrt.c

```diff
@@ -51,7 +51,7 @@ double invmag(double* __restrict__ A, int n) {
   return Q_rsqrt(sumsq);
 }
 
-
+//// (1) An interface that just happens to work because Q_rsqrt is simple enough.
```

Comment:
So the issue here isn't because of Q_sqrt, but rather it is because of the use of Q_sqrt within the function being differentated.

In particular, someones one may need the original primal return to compute the derivative of a function. e.g. if we did expr(Q_sqrt(x)). However, since we simply return Q_rsqrt (and the default semantics of autodiff without a flag is to just return the derivative, not the primal), we didn't need it.

In the second case, becuase it is stored into a duplicated, the same value must be stored in the derivative (and thus the primal return is required). If the variable A was marked dup_noneed, then the return-less version would've been fine.

The julia custom rule API goes into more details: https://enzyme.mit.edu/index.fcgi/julia/stable/generated/custom_rule/

That said this is very undocumented, and counter-intuitive.

@samuelpmish and I are working on trying to make a nicer c/c++ interface, including for custom derivatives so this is a great example to have.

Reply:
Ah, yes. I see. It is nice to know what legalCombinedForwardReverse does now.
For some reason, replacing the return with `return __builtin_exp(Q_sqrt(sumsq))` wasn't enough to trigger the assert, but additionally deleting the definition of `Q_sqrt` and replacing it with `extern float Q_sqrt(float)` was.